From fe81d57df288ec351e0b0aef3ae0350383d39a08 Mon Sep 17 00:00:00 2001
From: Shayne Chen <shayne.chen@mediatek.com>
Date: Tue, 19 Aug 2025 17:52:59 +0800
Subject: [PATCH 9/9] mtk: iw: support per-radio txpower setting

Usage:
iw phy phy0 set txpower fixed 1600 radio 0
iw phy phy0 set txpower fixed 1600 radio 1
iw phy phy0 set txpower fixed 1600 radio 2

Signed-off-by: Shayne Chen <shayne.chen@mediatek.com>
---
 phy.c | 29 ++++++++++++++++++++++-------
 1 file changed, 22 insertions(+), 7 deletions(-)

diff --git a/phy.c b/phy.c
index 03a7822..a1df3fd 100644
--- a/phy.c
+++ b/phy.c
@@ -706,10 +706,11 @@ static int handle_txpower(struct nl80211_state *state,
 			  enum id_input id)
 {
 	enum nl80211_tx_power_setting type;
+	char *endptr;
 	int mbm;
 
 	/* get the required args */
-	if (argc != 1 && argc != 2)
+	if (argc <= 0 && argc > 4)
 		return 1;
 
 	if (!strcmp(argv[0], "auto"))
@@ -725,26 +726,40 @@ static int handle_txpower(struct nl80211_state *state,
 
 	NLA_PUT_U32(msg, NL80211_ATTR_WIPHY_TX_POWER_SETTING, type);
 
+	argv++;
+	argc--;
+
 	if (type != NL80211_TX_POWER_AUTOMATIC) {
-		char *endptr;
-		if (argc != 2) {
+		if (argc != 1 && argc != 3) {
 			printf("Missing TX power level argument.\n");
 			return 2;
 		}
 
-		mbm = strtol(argv[1], &endptr, 10);
-		if (*endptr)
+		mbm = strtol(argv[0], &endptr, 10);
+		if (argc == 1 && *endptr)
 			return 2;
 		NLA_PUT_U32(msg, NL80211_ATTR_WIPHY_TX_POWER_LEVEL, mbm);
-	} else if (argc != 1)
+
+		argv++;
+		argc--;
+	} else if (argc != 0 && argc != 2)
 		return 1;
 
+	if (argc == 2 && strcmp("radio", argv[0]) == 0) {
+		unsigned int radio_idx = strtoul(argv[1], &endptr, 10);
+
+		if (*endptr)
+			return 2;
+
+		NLA_PUT_U8(msg, NL80211_ATTR_WIPHY_RADIO_INDEX, radio_idx);
+	}
+
 	return 0;
 
  nla_put_failure:
 	return -ENOBUFS;
 }
-COMMAND(set, txpower, "<auto|fixed|limit> [<tx power in mBm>]",
+COMMAND(set, txpower, "<auto|fixed|limit> [<tx power in mBm>] [radio <radio index>]",
 	NL80211_CMD_SET_WIPHY, 0, CIB_PHY, handle_txpower,
 	"Specify transmit power level and setting type.");
 COMMAND(set, txpower, "<auto|fixed|limit> [<tx power in mBm>]",
-- 
2.45.2

